sources/whippet/src/policy.rs

```diff
@@ -0,0 +1,923 @@
+        send_member: String,
+        #[serde(default)]
+        send_type: MessageType,
+        #[serde(default)]
+        send_path: String,
+        #[serde(default)]
+        send_broadcast: u32,
+        allow: bool,
+        #[serde(skip)]
+        priority: u64,
+    },
+    /// Receive rules allow services to receive messages from the bus
+    Receive {
+        #[serde(default)]
+        receive_sender: String,
+        #[serde(default)]
+        receive_path: String,
+        #[serde(default)]
+        receive_interface: String,
+        #[serde(default)]
+        receive_member: String,
+        #[serde(default)]
+        receive_type: MessageType,
+        #[serde(default)]
+        receive_broadcast: u32,
+        allow: bool,
+        #[serde(skip)]
+        priority: u64,
+    },
+}
+
+/// Represents the message type configured for the send and receive rules. The original dbus policy
+/// allows for MethodReturn and Error message types, however, in the dbus-launcher, rules with
+/// either message type are dropped and not serialized in the final policy
+///
+/// See:
+/// https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/launch/policy.c#L439
+#[derive(
+    Debug, Deserialize, Serialize_repr, Clone, PartialEq, Default, Copy, ZVariantType, ZVariantValue,
+)]
+#[repr(u32)]
+#[serde(rename_all = "kebab-case")]
+pub enum MessageType {
+    // Not really invalid, as it is a valid value, but this is the name given by the launcher
+    // and it is the default value used when it isn't present in a send or receive rule
+    //
+    // See:
+    // https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/dbus/protocol.h#L35
+    #[default]
+    Invalid = 0,
+    MethodCall = 1,
+    Signal = 4,
+}
+
+impl Policy {
+    /// Returns a new policy constructed from the provided file and additional policies. It loads the
+    /// policies in the passed directory in lexicographic order
+    #[must_use = "policy construction result must be checked"]
+    pub(crate) fn new<U>(
+        policy_file: impl AsRef<Path>,
+        policies_dir: impl AsRef<Path>,
+        mut username_resolver: U,
+    ) -> Result<Self>
+    where
+        U: UsernameResolver,
+    {
+        let policy_file = policy_file.as_ref();
+        let policies_dir = policies_dir.as_ref();
+        let mut all_policies: Vec<PathBuf> = vec![policy_file.into()];
+
+        let mut toml_files: Vec<_> = fs::read_dir(policies_dir)
+            .context(error::ReadPathSnafu {
+                what: policies_dir.display().to_string(),
+            })?
+            .collect::<std::io::Result<Vec<_>>>()
+            .context(error::CollectPathsSnafu {
+                what: policies_dir.display().to_string(),
+            })?
+            .into_iter()
+            .map(|entry| entry.path())
+            .filter(|path| path.extension().and_then(|s| s.to_str()) == Some("toml"))
+            .collect();
```

Comment:
nit: could factor out into a helper function to keep the main algorithm readable